src/DropdownMenu/DropdownMenu.tsx

```diff
@@ -0,0 +1,158 @@
+  }, [])
+
+  const onAnchorKeyDown = useCallback(
+    (event: React.KeyboardEvent<HTMLElement>) => {
+      if (!event.defaultPrevented) {
+        if (!open) {
+          if (event.key === 'ArrowDown' || event.key === 'ArrowUp') {
+            setFocusType('list')
+            setOpen(true)
+            event.preventDefault()
+          } else if (event.key === ' ' || event.key === 'Enter') {
+            setFocusType('anchor')
+            setOpen(true)
+            event.preventDefault()
+          }
+        } else if (focusType === 'anchor') {
+          if (['ArrowDown', 'ArrowUp', 'Tab', 'Enter'].includes(event.key)) {
+            setFocusType('list')
+            event.preventDefault()
+          } else if (event.key === 'Escape') {
+            onDismiss()
+            event.preventDefault()
+          }
+        }
+      }
+    },
+    [open, focusType, onDismiss]
+  )
+  const onAnchorClick = useCallback(
+    (event: React.MouseEvent<HTMLElement>) => {
+      if (!event.defaultPrevented && event.button === 0 && !open) {
+        setOpen(true)
+        setFocusType('anchor')
+      }
+    },
+    [open]
+  )
+
+  const {position} = useAnchoredPosition({anchorElementRef: anchorRef, floatingElementRef: overlayRef})
+
+  useFocusZone({containerRef: overlayRef, disabled: !open || focusType !== 'list' || !position})
+  useFocusTrap({containerRef: overlayRef, disabled: !open || focusType !== 'list' || !position})
+
+  return (
+    <>
+      {renderAnchor({
+        ref: anchorRef,
+        id: anchorId,
+        'aria-labelledby': anchorId,
+        'aria-haspopup': 'listbox',
+        children: selectedItem?.text ?? placeholder,
+        onClick: onAnchorClick,
+        onKeyDown: onAnchorKeyDown
+      })}
+      {open ? (
+        <Overlay
+          initialFocusRef={anchorRef}
+          returnFocusRef={anchorRef}
+          onClickOutside={onDismiss}
+          onEscape={onDismiss}
+          ref={updateOverlayRef}
+          {...position}
+        >
+          <List
+            {...listProps}
+            role="listbox"
+            renderItem={({onClick, onKeyDown, item, ...itemProps}) => {
+              const itemActivated = () => {
+                onChange?.(item === selectedItem ? undefined : item)
+                onDismiss()
+              }
+
+              return renderItem({
+                ...itemProps,
+                item,
+                role: 'option',
+                selected: item === selectedItem,
+                onClick: event => {
+                  itemActivated()
+                  onClick && onClick(event)
+                },
+                onKeyDown: event => {
+                  if (!event.defaultPrevented && [' ', 'Enter'].includes(event.key)) {
+                    itemActivated()
+                    // prevent "Enter" event from becoming a click on the anchor as overlay closes
+                    event.preventDefault()
+                  }
+                  onKeyDown && onKeyDown(event)
```

Comment:
I believe we can use the `?.` operator to clean this up a bit:
```suggestion
                  onClick?.(event)
                },
                onKeyDown: event => {
                  if (!event.defaultPrevented && [' ', 'Enter'].includes(event.key)) {
                    itemActivated()
                    // prevent "Enter" event from becoming a click on the anchor as overlay closes
                    event.preventDefault()
                  }
                  onKeyDown?.(event)
```